stackslib/src/net/inv/nakamoto.rs

```diff
@@ -233,3 +242,569 @@ impl InvGenerator {
+                &self.neighbor_address
+            );
+            return false;
+        }
+
+        // ask this neighbor for its inventory
+        true
+    }
+
+    /// Finish asking for inventories, and update inventory state.
+    /// Return Ok(true) if we learned something new
+    /// Return Ok(false) if not.
+    /// Return Err(..) on I/O errors
+    pub fn getnakamotoinv_try_finish(
+        &mut self,
+        network: &mut PeerNetwork,
+        reply: StacksMessage,
+    ) -> Result<bool, NetError> {
+        match reply.payload {
+            StacksMessageType::NakamotoInv(inv_data) => {
+                debug!(
+                    "{:?}: got NakamotoInv: {:?}",
+                    network.get_local_peer(),
+                    &inv_data
+                );
+                let ret =
+                    self.merge_tenure_inv(inv_data.tenures, inv_data.bitlen, self.reward_cycle());
+                self.next_reward_cycle();
+                return Ok(ret);
+            }
+            StacksMessageType::Nack(nack_data) => {
+                info!("{:?}: remote peer NACKed our GetNakamotoInv", network.get_local_peer();
+                      "error_code" => nack_data.error_code);
+                self.set_online(false);
+                return Ok(false);
+            }
+            _ => {
+                info!(
+                    "{:?}: got unexpected message from {:?}: {:?}",
+                    network.get_local_peer(),
+                    &self.neighbor_address,
+                    &reply
+                );
+                self.set_online(false);
+                return Err(NetError::ConnectionBroken);
+            }
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum NakamotoInvState {
+    GetNakamotoInvBegin,
+    GetNakamotoInvFinish,
+    Done,
+}
+
+/// Nakamoto inventory state machine
+pub struct NakamotoInvStateMachine<NC: NeighborComms> {
+    /// Communications links
+    pub(crate) comms: NC,
+    /// Nakamoto inventories we have
+    pub(crate) inventories: HashMap<NeighborAddress, NakamotoTenureInv>,
+    /// Reward cycle consensus hashes
+    reward_cycle_consensus_hashes: BTreeMap<u64, ConsensusHash>,
+}
+
+impl<NC: NeighborComms> NakamotoInvStateMachine<NC> {
+    pub fn new(comms: NC) -> Self {
+        Self {
+            comms,
+            inventories: HashMap::new(),
+            reward_cycle_consensus_hashes: BTreeMap::new(),
+        }
+    }
+
+    pub fn reset(&mut self) {
+        self.comms.reset();
+    }
+
+    /// Remove state for a particular neighbor
+    pub fn del_peer(&mut self, peer: &NeighborAddress) {
+        self.inventories.remove(peer);
+    }
+
+    /// Highest reward cycle learned
+    pub fn highest_reward_cycle(&self) -> u64 {
+        let mut highest_rc = 0;
+        for (_, inv) in self.inventories.iter() {
+            highest_rc = inv.highest_reward_cycle().max(highest_rc);
+        }
+        highest_rc
+    }
```

Comment:
Is this method used anywhere or is it only used in tests?

Reply:
Slightly clearer way to write this logic:

```suggestion
    pub fn highest_reward_cycle(&self) -> u64 {
        self.inventories.iter()
            .map(|(_, inv)| inv.highest_reward_cycle())
            .max()
            .unwrap_or(0)
    }
```

Reply:
It will be needed by the block downloader.